src/ppx/instrument.ml

```diff
@@ -32,26 +32,14 @@
    point is in [bisect_ppx.ml]. It's basically a PPX driver that registers only
    this instrumenter with itself, using [register.ml], and then runs it. *)
 
+open Ppxlib
```

Comment:
Is it possible and reasonable to avoid this file-wide `open`?

Reply:
That should be possible. But I've just tried and I'm not sure to which extend it would be "reasonable": if we don't open Ppxlib file-wide, OCaml would have a hard time inferring types correctly. E.g. [here](https://github.com/aantron/bisect_ppx/pull/327/files#diff-4bdd650c927dc010be2484e34719df1b48a1c6debc9f8bb81e97f822ae3d7de6R149) it would infer that `e` has type `Parsetree.expression` whereas we want it to be of type `Ppxlib.Parsetree.expression`. And similar in a lot of places. Setting `module Parsetree = Ppxlib.Parstree` wouldn't help in that regard. So apart from overwriting several modules, we'd have to specify a lot of types throughout the whole file.

So in summary, I think it would be possible but quite some work. Do you think it's necessary?

Reply:
I think it's fine to leave it how it is.

Reply:
Actually, this is inaccurate:

> it would be possible but quite some work

I just tried to do it myself and got a first version working in less than 2 minutes.

> E.g. [here](https://github.com/aantron/bisect_ppx/pull/327/files#diff-4bdd650c927dc010be2484e34719df1b48a1c6debc9f8bb81e97f822ae3d7de6R149) it would infer that `e` has type `Parsetree.expression` whereas we want it to be of type `Ppxlib.Parsetree.expression`. And similar in a lot of places. Setting `module Parsetree = Ppxlib.Parstree` wouldn't help in that regard. So apart from overwriting several modules, we'd have to specify a lot of types throughout the whole file.

I did exactly what you suggested and it worked. I added

```ocaml
module Parsetree = Ppxlib.Parsetree
module Ast_builder = Ppxlib.Ast_builder
```

and places where metaquot was used needed

```
let open Parsetree in
```

before them.

Some uses of `Nolabel` (which this PR un-qualified) needed to be re-qualified. Overall it was extremely easy to rush through it and get it building. Could you please quickly do a proper version of it?

Reply:
Opening `Parsetree` locally before metaquot uses (instead of type specifications) is a good idea, thanks! And you were right: it has been far less work than I thought.

I've opened `Ocaml_shadow` in the ppx dune file now to make sure that people don't accidentally use compiler-libs in the future. Opening Ppxlib would have given the same protection.

Reply:
> I've opened `Ocaml_shadow` in the ppx dune file now to make sure that people don't accidentally use compiler-libs in the future.

Nice, thanks.